pkg/agent/controller/noderoute/node_route_controller.go

```diff
@@ -144,35 +147,178 @@ func (c *Controller) enqueueNode(obj interface{}) {
 	}
 }
 
+// listRoutes reads all the routes for the gateway and returns them as a map with the destination
+// subnet as the key.
+func (c *Controller) listRoutes() (routes map[string]*netlink.Route, err error) {
+	routes = make(map[string]*netlink.Route)
+	routeList, err := netlink.RouteList(c.gatewayLink, netlink.FAMILY_V4)
+	if err != nil {
+		return routes, err
+	}
+	for _, route := range routeList {
+		routes[route.Dst.String()] = &route
+	}
+	return routes, nil
+}
+
+// removeStaleGatewayRoutes removes all the gateway routes which no longer correspond to a Node in
+// the cluster. If the antrea agent restarts and Nodes have left the cluster, this function will
+// take care of removing routes which are no longer valid.
+func (c *Controller) removeStaleGatewayRoutes() error {
+	nodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("error when listing Nodes: %v", err)
+	}
+
+	routes, err := c.listRoutes()
+	if err != nil {
+		return fmt.Errorf("error when listing existing gateway routes: %v", err)
+	}
+
+	for _, node := range nodes {
+		_, podCIDR, err := net.ParseCIDR(node.Spec.PodCIDR)
+		if err != nil {
+			return fmt.Errorf("error when parsing Pod CIDR for Node %s: %v", node.Name, err)
+		}
+		podCIDRStr := podCIDR.String()
+		// remove from map of routes: at the end of the loop the remaining routes will be
+		// the stale ones (the ones we want to delete).
+		// we iterate over all current Nodes, including the Node on which this agent is
+		// running, so the route to local Pods will be removed from the map as well.
+		delete(routes, podCIDRStr)
+	}
+
+	// remove all remaining routes in the map: these routes do not match any Node in the
+	// cluster.
+	klog.V(4).Infof("Need to delete %d routes", len(routes))
+	for _, route := range routes {
+		if err := netlink.RouteDel(route); err != nil {
+			klog.Errorf("error when deleting route '%v': %v", route, err)
+		}
+	}
+
+	return nil
+}
+
+// removeStaleTunnelPorts removes all the tunnel ports which no longer correspond to a Node in the
+// cluster. If the antrea agent restarts and Nodes have left the cluster, this function will take
+// care of removing tunnel ports which are no longer valid. If the tunnel port configuration has
+// changed, the tunnel port will also be deleted (the controller loop will later take care of
+// re-creating the port with the correct configuration).
+func (c *Controller) removeStaleTunnelPorts() error {
+	nodes, err := c.nodeLister.List(labels.Everything())
```

Comment:
Does this always call to K8s API? If yes, probably makes sense to share a single call with removeStaleGatewayRoutes.

Reply:
I'm pretty sure it doesn't. It uses the local client cache.